File: context_os_runtime/state.py
from __future__ import annotations

from enum import StrEnum


class SessionState(StrEnum):
    BOUND = "BOUND"
    IDLE = "IDLE"
    PLANNED = "PLANNED"
    AWAITING_APPROVAL = "AWAITING_APPROVAL"
    EXECUTING = "EXECUTING"
    EXECUTED = "EXECUTED"
    VERIFIED = "VERIFIED"
    REVIEWED = "REVIEWED"
    COMPLETE = "COMPLETE"


_ALLOWED = {
    SessionState.BOUND: {SessionState.IDLE},
    SessionState.IDLE: {SessionState.PLANNED},
    SessionState.PLANNED: {SessionState.AWAITING_APPROVAL, SessionState.EXECUTING},
    SessionState.AWAITING_APPROVAL: {SessionState.IDLE, SessionState.EXECUTING},
    SessionState.EXECUTING: {SessionState.EXECUTED},
    SessionState.EXECUTED: {SessionState.VERIFIED},
    SessionState.VERIFIED: {SessionState.REVIEWED, SessionState.COMPLETE},
    SessionState.REVIEWED: {SessionState.COMPLETE},
    SessionState.COMPLETE: set(),
}


def transition(current: SessionState, target: SessionState) -> SessionState:
    if target not in _ALLOWED[current]:
        raise ValueError(f"invalid transition: {current} -> {target}")
    return target

File: src/core/state.ts
export const SessionState = {
  BOUND: 'BOUND',
  IDLE: 'IDLE',
  PLANNED: 'PLANNED',
  AWAITING_APPROVAL: 'AWAITING_APPROVAL',
  EXECUTING: 'EXECUTING',
  EXECUTED: 'EXECUTED',
  VERIFIED: 'VERIFIED',
  REVIEWED: 'REVIEWED',
  COMPLETE: 'COMPLETE',
} as const;
export type SessionState = (typeof SessionState)[keyof typeof SessionState];

const ALLOWED: Record<SessionState, ReadonlySet<SessionState>> = {
  BOUND: new Set([SessionState.IDLE]),
  IDLE: new Set([SessionState.PLANNED]),
  PLANNED: new Set([SessionState.AWAITING_APPROVAL, SessionState.EXECUTING]),
  AWAITING_APPROVAL: new Set([SessionState.IDLE, SessionState.EXECUTING]),
  EXECUTING: new Set([SessionState.EXECUTED]),
  EXECUTED: new Set([SessionState.VERIFIED]),
  VERIFIED: new Set([SessionState.REVIEWED, SessionState.COMPLETE]),
  REVIEWED: new Set([SessionState.COMPLETE]),
  COMPLETE: new Set(),
};

export function transition(current: SessionState, target: SessionState): SessionState {
  if (!ALLOWED[current].has(target)) {
    throw new Error(`invalid transition: ${current} -> ${target}`);
  }
  return target;
}
